torch/_inductor/ir.py

```diff
@@ -5261,3 +5262,97 @@ def codegen_collective(self, wrapper, output_name, input_names):
             f"group={output_name}_pg, "
             "async_op=True)"
         )
+
+
+class AllToAllSingle(OutOfPlaceCollectiveKernel):
+    count = itertools.count()
+
+    def __init__(
+        self, layout, inputs, outputs, constant_args,
+        inputs_allow_none,
+    ):
+        super().__init__(layout, inputs, outputs, constant_args)
+        self.inputs_allow_none = inputs_allow_none
+
+    @classmethod
+    def create(
+        cls,
+        x: "TensorBox",
+        output_split_sizes: Optional["TensorBox"],
+        input_split_sizes: Optional["TensorBox"],
+        tag: str,
+        ranks: List[int],
+        group_size: int,
+    ):
+        x_realized = cls.realize_input(x)
+        inputs_allow_none = []
+        output_split_sizes_realized = None
+        if output_split_sizes is not None:
+            output_split_sizes_realized = cls.realize_input(output_split_sizes)
+            # NCCL/GLOO alltoall_base also assumes output_split_sizes to be int64,
+            # so hardcoding int64 here should be okay.
+            V.graph.name_to_coerced_dtype[output_split_sizes_realized.data.name] = torch.int64
+        input_split_sizes_realized = None
+        if input_split_sizes is not None:
+            input_split_sizes_realized = cls.realize_input(input_split_sizes)
+            # NCCL/GLOO alltoall_base also assumes input_split_sizes to be int64,
+            # so hardcoding int64 here should be okay.
+            V.graph.name_to_coerced_dtype[input_split_sizes_realized.data.name] = torch.int64
+        inputs_allow_none = [x_realized, output_split_sizes_realized, input_split_sizes_realized]
+        inputs = [inp for inp in inputs_allow_none if inp is not None]
+
+        outputs = []
+        new_size = x_realized.get_size()
+        output_split_sizes_sum_s = None
+        if output_split_sizes is not None:
+            output_split_sizes_sum_s = V.graph.current_node.meta['val'].size()[0]
+            new_size[0] = output_split_sizes_sum_s
+            # FlexibleLayout ctor used below needs to know which sizevar in output size
+            # is okay to skip the `isinstance(s, (Expr, int))` check for.
+            V.graph.data_dep_symbol_to_expr_str[str(output_split_sizes_sum_s)] = None
```

Comment:
ok so, is it correct that you use this dict as a 3 step pipeline?
1) put a key = none to signal that key needs to be populated
2) in codegen, update the key to an expression which..i guess you couldn't have created at step 1 because... idk, maybe the output names aren't known yet?
3) in wrapper codegen, utilize the expression from (2) to create the sizevar that will be used to alloc the output that is referenced by codegen in (2)?

it's pretty convoluted but i can see that it would work.  I just wonder if there would be a way to make it cleaner (perhaps in a later PR)?

(oh, and it'd be good to put a comment somewhere that explains this 3 steps)

Reply:
thanks - after a few iterations, no more need for this 3-step pipeline (yay!)